x/stake/handler_sidechain.go

```diff
@@ -476,6 +480,102 @@ func handleMsgSideChainUndelegate(ctx sdk.Context, msg MsgSideChainUndelegate, k
 	return sdk.Result{Data: finishTime, Tags: tags, Events: events}
 }
 
+func handleMsgSideChainStakeMigration(ctx sdk.Context, msg MsgSideChainStakeMigration, k keeper.Keeper) sdk.Result {
+	if scCtx, err := k.ScKeeper.PrepareCtxForSideChain(ctx, k.DestChainName); err != nil {
+		return ErrInvalidSideChainId(k.Codespace()).Result()
+	} else {
+		ctx = scCtx
+	}
+
+	denom := k.BondDenom(ctx)
+	if msg.Amount.Denom != denom {
+		return ErrBadDenom(k.Codespace()).Result()
+	}
+
+	shares, sdkErr := k.ValidateUnbondAmount(ctx, msg.RefundAddress, msg.Validator, msg.Amount.Amount)
+	if sdkErr != nil {
+		return sdkErr.Result()
+	}
+
+	// unbond immediately
+	ubd, sdkErr := k.BeginUnbonding(ctx, msg.RefundAddress, msg.Validator, shares, false)
+	if sdkErr != nil {
+		return sdkErr.Result()
+	}
+	ubd, events, sdkErr := k.CompleteUnbonding(ctx, ubd.DelegatorAddr, ubd.ValidatorAddr)
+	if sdkErr != nil {
+		return sdkErr.Result()
+	}
+
+	// send coins to pegAccount
+	_, sdkErr = k.BankKeeper.SendCoins(ctx, msg.RefundAddress, sdk.PegAccount, sdk.Coins{ubd.Balance})
+	if sdkErr != nil {
+		return sdkErr.Result()
+	}
+
+	// send cross-chain package
+	bscAmount := bsc.ConvertBCAmountToBSCAmount(ubd.Balance.Amount)
+	stakeMigrationSynPackage := types.StakeMigrationSynPackage{
+		OperatorAddress:  msg.OperatorAddress,
+		DelegatorAddress: msg.DelegatorAddress,
+		RefundAddress:    msg.RefundAddress,
+		Amount:           bscAmount,
+	}
+
+	encodedPackage, err := rlp.EncodeToBytes(stakeMigrationSynPackage)
+	if err != nil {
+		return sdk.ErrInternal("encode stake migration package error").Result()
+	}
+
+	relayFee := bsc.ConvertBCAmountToBSCAmount(types.StakeMigrationRelayFee)
+	sendSeq, sdkErr := k.IbcKeeper.CreateRawIBCPackageByIdWithFee(ctx, k.DestChainId, types.StakeMigrationChannelID, sdk.SynCrossChainPackageType,
+		encodedPackage, *relayFee)
+	if sdkErr != nil {
+		return sdkErr.Result()
+	}
+
+	if k.PbsbServer != nil && ctx.IsDeliverTx() {
+		uEvent := types.ChainUndelegateEvent{
+			UndelegateEvent: types.UndelegateEvent{
+				StakeEvent: types.StakeEvent{
+					IsFromTx: true,
+				},
+				Delegator: msg.RefundAddress,
+				Validator: msg.Validator,
+				Amount:    msg.Amount.Amount,
+				Denom:     msg.Amount.Denom,
+				TxHash:    ctx.Value(baseapp.TxHashKey).(string),
+			},
+			ChainId: k.DestChainName,
+		}
+		k.PbsbServer.Publish(uEvent)
+
+		ctEvent := pubsub.CrossTransferEvent{
```

Comment:
who will consume this message, is it useful ? @forcodedancing

Reply:
Reconcilation?

Reply:
Yes, I think it is used by Recon.